Navigator: stabilize and export APIs (#64613)
WordPress/gutenberg: packages/components/src/navigator/README.md

# `Navigator`

> `Navigator` is a collection components that allow rendering nested views/panels/menus (via the [`Navigator.Screen` component](/packages/components/src/navigator/navigator-screen/README.md)) and navigate between them (via the [`Navigator.Button`](/packages/components/src/navigator/navigator-button/README.md) and [`Navigator.BackButton`](/packages/components/src/navigator/navigator-back-button/README.md) components).

Reviewer: Didn't we remove these READMEs? 

```suggestion
`Navigator` is a collection of components that allow rendering nested views/panels/menus (via the `Navigator.Screen`) and navigate between them (via the `Navigator.Button` and `Navigator.BackButton` components).
```

Author: 🤦

## Usage

```jsx
import { Navigator } from '@wordpress/components';

const MyNavigation = () => (
	<Navigator initialPath="/">
		<Navigator.Screen path="/">
			<p>This is the home screen.</p>
			<Navigator.Button path="/child">
				Navigate to child screen.
			</Navigator.Button>
		</Navigator.Screen>
		<Navigator.Screen path="/child">
			<p>This is the child screen.</p>
			<Navigator.BackButton>Go back</Navigator.BackButton>
		</Navigator.Screen>
	</Navigator>
);
```

### Hierarchical `path`s

`Navigator` assumes that screens are organized hierarchically according to their `path`, which should follow a URL-like scheme where each path segment starts with and is separated by the `/` character.

> `Navigator` will treat "back" navigations as going to the parent screen — it is therefore responsibility of the consumer of the component to create the correct screen hierarchy.

Reviewer: ```suggestion
`Navigator` will treat "back" navigations as going to the parent screen — it is, therefore, the responsibility of the consumer of the component to create the correct screen hierarchy.
```

### Height and animations

Due to how `Navigator.Screen` animations work, it is recommended that the `Navigator` component is assigned a `height` to prevent some potential UI jumps while moving across screens.

### Individual components

`Navigator` is comprised of four individual components:

-   `Navigator`: a wrapper component and context provider. It holds the main logic for hiding and showing screens.
-   `Navigator.Screen`: represents a single view/screen/panel;
-   `Navigator.Button`: renders a button that allows navigating to a different `Navigator.Screen`;
-   `Navigator.BackButton`: renders a button that allows navigating to the parent `Navigator.Screen` (see the section above about hierarchical paths).

For advanced usages, consumers can use the `useNavigator` hook.

#### `Navigator`

##### Props

###### `initialPath`: `string`

The initial active path.

-   Required: Yes

###### `children`: `string`

The children elements.

-   Required: Yes

#### `Navigator.Screen`

###### `path`: `string`

> The screen&quot;s path, matched against the current path stored in the navigator.

Reviewer: ```suggestion
The screen's path, matched against the current path stored in the navigator.
```

`Navigator` assumes that screens are organized hierarchically according to their `path`, which should follow a URL-like scheme where each path segment starts with and is separated by the `/` character.

> `Navigator` will treat "back" navigations as going to the parent screen — it is therefore responsibility of the consumer of the component to create the correct screen hierarchy.

Reviewer: ```suggestion
`Navigator` will treat "back" navigations as going to the parent screen — it is, therefore, the responsibility of the consumer of the component to create the correct screen hierarchy.
```

###### `useNavigator`

You can retrieve a `navigator` instance by using the `useNavigator` hook.

The `navigator` instance has a few properties:

###### `goTo`: `( path: string, options: NavigateOptions ) => void`

The `goTo` function allows navigating to a given path. The second argument can augment the navigation operations with different options.

The available options are:

-   `focusTargetSelector`: `string`. An optional property used to specify the CSS selector used to restore focus on the matching element when navigating back;
-   `isBack`: `boolean`. An optional property used to specify whether the navigation should be considered as backwards (thus enabling focus restoration when possible, and causing the animation to be backwards too);
-   `skipFocus`: `boolean`. An optional property used to opt out of `Navigator`'s focus management, useful when the consumer of the component wants to manage focus themselves;

###### `goBack`: `( path: string, options: NavigateOptions ) => void`

The `goBack` function allows navigating to the parent screen. Parent/child navigation only works if the paths you define are hierarchical (see note above).

> When a match is not found, the function will try to recursively navigate the path hierarchy until a matching screen (or the root `/`) are found.

Reviewer: ```suggestion
When a match is not found, the function will try to recursively navigate the path hierarchy until a matching screen (or the root `/`) is found.
```

The available options are the same as for the `goTo` method, except for the `isBack` property, which is not available for the `goBack` method.

###### `location`: `NavigatorLocation`

> The `location` object represent the current location, and has a few properties:

Reviewer: ```suggestion
The `location` object represents the current location, and has a few properties:
```
